### PR TITLE
add video_link field + tests

### PR DESCRIPTION
**Pull request by Elkorm98**
https://trello.com/c/lHUW3Mht/27-be-add-videolink-field-to-recipes-return-it-in-response-l

### CODE REVIEW

File: app/serializers/recipe_serializer.rb [hunk 1]
```diff
@@ -1,3 +1,10 @@
 class RecipeSerializer < ActiveModel::Serializer
-  attributes :id, :name, :content
+  attributes :id, :name, :content, :video_link
+  UNKNOWN = 'unknown'
```
Comment (h4b00) [inline app/serializers/recipe_serializer.rb:3]: .freeze

File: spec/controllers/api/v1/recipes_controller_spec.rb [hunk 2]
```diff
@@ -63,33 +68,51 @@
...
+        expect do
+          put :update, params: { id: recipe.id, recipe: { video_link: invalid_new_video_link } }
+        end.not_to change { recipe.reload.video_link }
+      end
     end
   end
 
   describe '[DELETE] #destroy' do
     let!(:recipe) { create(:recipe) }
+    let(:recipeJSON) { JSON.parse(response.body) }
```
Comment (h4b00) [inline spec/controllers/api/v1/recipes_controller_spec.rb:105]: don't use camel syntax use ``_

File: app/models/recipe.rb [hunk 3]
```diff
@@ -1,3 +1,4 @@
 class Recipe < ApplicationRecord
   validates :name, presence: true
+  validates :video_link, url: { no_local: true, schemes: ['https'] }, length: { in: 10..100 }, allow_blank: true
```
Comment (sebastiansito) [inline app/models/recipe.rb:3]: Do not use hardcoded numbers or other values like that. Everyone will immediately wonder why 10-100 not for example 20-200. Code should be written in a way that the intent is clear
```suggestion
  URL_MIN_LENGTH = 10
  URL_MAX_LENGTH = 100
  
  validates :video_link, url: { no_local: true, schemes: ['https'] }, length: { in: URL_MIN_LENGTH..URL_MAX_LENGTH }, allow_blank: true
```
When you do that you'll see that URLs cannot be capped at 100 characters. They are 2048 max (https://www.google.com/search?q=url+max+length&oq=url+max+le&aqs=chrome.0.0i512j69i57j0i512l4j0i22i30l4.1280j0j7&sourceid=chrome&ie=UTF-8).
Also why do we check for URL length? Is URL validator not enough?

File: app/serializers/recipe_serializer.rb (hunk 1 above)
Comment (sebastiansito) [inline app/serializers/recipe_serializer.rb:3]: Consts should be put at the top of the file separated by empty newline from the rest of the code

**Review by h4b00 — APPROVED** [review]
Fix just this camelCase.
The rest looks good!

File: spec/controllers/api/v1/recipes_controller_spec.rb [hunk 4]
```diff
@@ -29,7 +30,8 @@
       {
         recipe: {
           name: 'Leczo',
-          content: 'Very good dish'
+          content: 'Very good dish',
+          video_link: 'https://www.youtube.com/watch?v=60KwNYk09BU'
```
Comment (sebastiansito) [inline spec/controllers/api/v1/recipes_controller_spec.rb:34]: Use `let` so you don't have to type the same URL all over again.
```suggestion
          let(:video_link) { 'https://www.youtube.com/watch?v=60KwNYk09BU' }
          # ...
          video_link: video_link
```

**Review by h4b00 — CHANGES_REQUESTED** [review]
missing migration?

File: spec/controllers/api/v1/recipes_controller_spec.rb [hunk 5]
```diff
@@ -45,7 +47,7 @@
     context 'with invalid params' do
       it 'doesn\'t create recipe' do
         expect do
-          post :create, params: { recipe: { name: '', content: '' } }
+          post :create, params: { recipe: { name: '', content: '', video_link: 'http://www.youtube.com/watch?v=60KwNYk09BU' } }
```
Comment (sebastiansito) [inline spec/controllers/api/v1/recipes_controller_spec.rb:50]: If you had `let` you could use it here

File: spec/controllers/api/v1/recipes_controller_spec.rb [hunk 6]
```diff
@@ -54,7 +56,10 @@
   describe '[PUT] #update' do
     let(:old_name) { 'Rosol' }
     let(:new_name) { 'Pierogi' }
-    let(:recipe) { create(:recipe, name: old_name) }
+    let(:old_video_link) { 'https://www.youtube.com/watch?v=60KwNYk09BU' }
+    let(:new_video_link) { 'https://www.youtube.com/watch?v=74re1sJJFtg' }
+    let(:recipe) { create(:recipe, name: old_name, video_link: old_video_link) }
+    let(:recipeJSON) { JSON.parse(response.body) }
```
Comment (sebastiansito) [inline spec/controllers/api/v1/recipes_controller_spec.rb:62]: Ruby uses snake case not camel case to name variables
```suggestion
    let(:recipe_json) { JSON.parse(response.body) }
```

File: spec/controllers/api/v1/recipes_controller_spec.rb [hunk 7]
```diff
@@ -63,33 +68,51 @@
         end.to change { recipe.reload.name }.from(old_name).to(new_name)
       end
 
+      it 'updates video_link' do
+        expect do
```
Comment (sebastiansito) [inline spec/controllers/api/v1/recipes_controller_spec.rb:72]: Good 👏

File: spec/controllers/api/v1/recipes_controller_spec.rb [hunk 8]
```diff
@@ -63,33 +68,51 @@
...
     it 'returns destroyed object' do
       delete :destroy, params: { id: recipe.id }
-      expect(JSON.parse(response.body)['name']).to eq(recipe.name)
+
+      expect { recipeJSON['name'].to eq(new_name) }
+      expect { recipeJSON['video_link'].to eq(new_video_link) }
     end
   end
-end
+end
```
Comment (sebastiansito) [inline spec/controllers/api/v1/recipes_controller_spec.rb:118]: Fix your code editor to add new lines automatically

File: spec/models/recipe_spec.rb [hunk 9]
```diff
@@ -3,5 +3,18 @@
...
+      it { is_expected.to validate_length_of(:video_link).is_at_least(10).is_at_most(100) }
+      it { is_expected.to allow_value('').for(:video_link) }
+    end
+    
+    context "video_link with invalid data" do  
+      it { is_expected.not_to allow_value('https://localhost:3000/').for(:video_link) }
+      it { is_expected.not_to allow_value('http://localhost:3000/').for(:video_link) }
+      it { is_expected.not_to validate_length_of(:video_link).is_at_most(9) }
+      it { is_expected.not_to validate_length_of(:video_link).is_at_least(91) }
+    end
```
Comment (sebastiansito) [inline spec/models/recipe_spec.rb:18]: `validate_url` gem comes with RSpec matcher: https://github.com/perfectline/validates_url#with-rspec so you don't have to test that manually

File: spec/serializers/recipe_serializer_spec.rb [hunk 10]
```diff
@@ -7,4 +7,18 @@
...
   it { is_expected.to include(content: recipe.content) }
+
+  describe 'video_link' do
+    context 'when video_link is nil' do
+      let(:recipe) { build(:recipe, video_link: nil) }
+
+      it { is_expected.to include(video_link: 'unknown') }
+    end
+  
+  context 'when vide_link is present' do
```
Comment (sebastiansito) [inline spec/serializers/recipe_serializer_spec.rb:18]: Indentation is messed up here

**Review by sebastiansito — CHANGES_REQUESTED** [review]

File: spec/serializers/recipe_serializer_spec.rb (hunk 10 above)
Comment (sebastiansito) [reply spec/serializers/recipe_serializer_spec.rb:18]: Also, typo `vide_link`

File: spec/controllers/api/v1/recipes_controller_spec.rb (hunk 5 above)
Comment (Elkorm98) [reply spec/controllers/api/v1/recipes_controller_spec.rb:50]: link used there is unvalidated

File: app/serializers/recipe_serializer.rb [hunk 11]
```diff
@@ -1,3 +1,11 @@
+UNKNOWN = 'unknown'.freeze
```
Comment (h4b00) [inline app/serializers/recipe_serializer.rb:1]: this const should be inside the class.

File: db/schema.rb [hunk 12]
```diff
@@ -20,6 +20,8 @@
     t.text "content"
     t.datetime "created_at", precision: 6, null: false
     t.datetime "updated_at", precision: 6, null: false
+    t.decimal "price", precision: 8, scale: 2
```
Comment (h4b00) [inline db/schema.rb:23]: why do you have a price here? you should have just `video_link`

File: spec/controllers/api/v1/recipes_controller_spec.rb (hunk 4 above)
Comment (h4b00) [reply spec/controllers/api/v1/recipes_controller_spec.rb:34]: @Elkorm98 please do as Seb wrote above ☝️

File: Gemfile [hunk 13]
```diff
@@ -24,6 +24,8 @@ gem 'bootsnap', '>= 1.4.4', require: false
 
 # Use Rack CORS for handling Cross-Origin Resource Sharing (CORS), making cross-origin AJAX possible
 # gem 'rack-cors'
+# url validator
+gem 'validate_url', '~> 1.0.13'
```
Comment (h4b00) [inline Gemfile:28]:
```suggestion

# url validator
gem 'validate_url', '~> 1.0.13'
```

**Review by h4b00 — CHANGES_REQUESTED** [review]

File: app/serializers/recipe_serializer.rb [hunk 14]
```diff
@@ -1,7 +1,7 @@
 class RecipeSerializer < ActiveModel::Serializer
-  UNKNOWN = 'unknown'.freeze
+  attributes :id, :name, :content, :price, :level, :video_link
 
-  attributes :id, :name, :content, :price, :level
+  UNKNOWN = 'unknown'.freeze
```
Comment (h4b00) [inline app/serializers/recipe_serializer.rb:4]: Consts should be first.

File: db/schema.rb [hunk 15]
```diff
@@ -21,19 +21,8 @@
...
-    t.string "email", default: "", null: false
-    t.string "encrypted_password", default: "", null: false
-    t.string "reset_password_token"
-    t.datetime "reset_password_sent_at"
-    t.datetime "remember_created_at"
-    t.datetime "created_at", precision: 6, null: false
-    t.datetime "updated_at", precision: 6, null: false
-    t.index ["email"], name: "index_users_on_email", unique: true
-    t.index ["reset_password_token"], name: "index_users_on_reset_password_token", unique: true
-  end
```
Comment (h4b00) [inline db/schema.rb:37]: Why do you want to delete it?

**Review by h4b00 — APPROVED** [review]

**Review by h4b00 — CHANGES_REQUESTED** [review]
Fix schema.db

File: db/schema.rb (hunk 15 above)
Comment (morales-cmd) [reply db/schema.rb]: its actually my fault. I accidentaly merged schema looking like this into master, although there is no migration creating user table. I must have forgotten to make rails db:migrate before merging. My bad, im sorry. Can it be fixed?

Comment (sebastiansito) [reply db/schema.rb]: You have to run `rails db:migrate` which will re-create the schema and then commit changes

Comment (morales-cmd) [reply db/schema.rb]: should i do it from my branch that was merged or can łukasz just do it from this branch?

Comment (Elkorm98) [reply db/schema.rb]: i think i fixed my schema.rb and its looks like there is no conflicts with master

**Pull request merged**